internal/resource/resource_service.go

```diff
@@ -160,6 +177,74 @@ func (r *ResourceService) ApplyConfig(ctx context.Context, instanceID string) er
 	return nil
 }
 
+func (r *ResourceService) GetUpstreams(ctx context.Context, instance *mpi.Instance,
+	upstream string,
+) ([]client.UpstreamServer, error) {
+	plusClient, err := r.createPlusClient(instance)
+	if err != nil {
+		slog.ErrorContext(ctx, "failed to create plus client ", "err", err)
+		return nil, err
+	}
+
+	servers, getServersErr := plusClient.GetHTTPServers(ctx, upstream)
+
+	return servers, createPlusAPIError(getServersErr)
+}
+
+// max number of returns from function is 3
+// nolint: revive
+func (r *ResourceService) UpdateHTTPUpstreams(ctx context.Context, instance *mpi.Instance, upstream string,
+	upstreams []*structpb.Struct,
+) (added, updated, deleted []client.UpstreamServer, err error) {
+	plusClient, err := r.createPlusClient(instance)
+	if err != nil {
+		slog.ErrorContext(ctx, "failed to create plus client ", "err", err)
+		return nil, nil, nil, err
+	}
+
+	servers := convertToUpstreamServer(upstreams)
+
+	return plusClient.UpdateHTTPServers(ctx, upstream, servers)
```

Comment:
Need to call `createPlusAPIError` before returning the error from the `plusClient.UpdateHTTPServers(ctx, upstream, servers)` function